Very easy to b0rk retention lease when subscribing (logical replication)

### CrateDB version

4.8.0

### CrateDB setup information

Number of nodes on publisher: 2


### Steps to Reproduce

Create a subscription:

```sql
create subscription iss_position connection 'crate://crate-public.eks1.eu-west-1.aws.cratedb.net?mode=pg_tunnel&user=public&password=public' publication iss_position;
```

This should replicate a iss_position table that gets a new write (row) every 10s.

Wait for a bit, then delete subscription, drop table, recreate subscription.

### Expected Result

Should work as expected.

### Actual Result

Getting something like:
```
org.elasticsearch.indices.recovery.RecoveryFailedException: [iss_position][0]: Recovery failed on {Clariden}{Cwd3VxbmQb2X6lSbltT2Iw}{0l97FiXOQVizTqBLU8CFuQ}{172.17.0.2}{172.17.0.2:4300}{http_address=172.17.0.2:4200}
...
Caused by: org.elasticsearch.index.shard.IndexShardRecoveryException: failed recovery
	... 58 more
Caused by: org.elasticsearch.index.snapshots.IndexShardRestoreFailedException: restore failed
	... 56 more
Caused by: org.elasticsearch.index.seqno.RetentionLeaseInvalidRetainingSeqNoException: the current retention lease with [logical_replication:crate:[iss_position][0]] is retaining a higher sequence number [12] than the new retaining sequence number [0] from [logical_replication:crate]
```

--- a/docs/appendices/release-notes/unreleased.rst
+++ b/docs/appendices/release-notes/unreleased.rst
@@ -112,3 +112,6 @@ Fixes
 - Fixed an issue with client caching which lead to authentication error when
   creating a subscription with bad credentials and ``pg_tunnel`` followed by
   re-creating it second time with the same name and valid credentials.
+
+- Fixed an issue resulting in a broken subscription when a subscription is
+  dropped and re-created within a short period of time.

--- a/server/src/main/java/io/crate/replication/logical/ShardReplicationChangesTracker.java
+++ b/server/src/main/java/io/crate/replication/logical/ShardReplicationChangesTracker.java
@@ -56,6 +56,7 @@ import io.crate.execution.support.RetryListener;
 import io.crate.execution.support.RetryRunnable;
 import io.crate.replication.logical.action.ReplayChangesAction;
 import io.crate.replication.logical.action.ShardChangesAction;
+import io.crate.replication.logical.exceptions.InvalidShardEngineException;
 import io.crate.replication.logical.seqno.RetentionLeaseHelper;
 
 /**
@@ -175,6 +176,10 @@ public class ShardReplicationChangesTracker implements Closeable {
                         );
                     }
                     updateBatchFetched(false, fromSeqNo, toSeqNo, fromSeqNo - 1, -1);
+                } else if (t instanceof InvalidShardEngineException) {
+                    if (LOGGER.isDebugEnabled()) {
+                        LOGGER.debug("Shard is not accepting replayed changes, engine changed", t);
+                    }
                 } else {
                     LOGGER.warn(
                         "[{}] Error during tracking of upstream shard changes for subscription '{}'. Tracking stopped: {}",

--- a/server/src/main/java/io/crate/replication/logical/ShardReplicationChangesTracker.java
+++ b/server/src/main/java/io/crate/replication/logical/ShardReplicationChangesTracker.java
@@ -139,7 +139,10 @@ public class ShardReplicationChangesTracker implements Closeable {
         long fromSeqNo = rangeToFetch.fromSeqNo();
         long toSeqNo = rangeToFetch.toSeqNo();
 
-        var futureClient = shardReplicationService.getRemoteClusterClient(shardId.getIndex());
+        var futureClient = shardReplicationService.getRemoteClusterClient(
+            shardId.getIndex(),
+            subscriptionName
+        );
         var getPendingChangesRequest = new ShardChangesAction.Request(shardId, fromSeqNo, toSeqNo);
         var futurePendingChanges = futureClient.thenCompose(remoteClient -> {
             if (LOGGER.isDebugEnabled()) {
@@ -161,7 +164,7 @@ public class ShardReplicationChangesTracker implements Closeable {
                 updateBatchFetched(true, fromSeqNo, toSeqNo, lastSeqNo, pendingChanges.lastSyncedGlobalCheckpoint());
             } else {
                 var t = SQLExceptions.unwrap(e);
-                if (SQLExceptions.maybeTemporary(t)) {
+                if (!closed && SQLExceptions.maybeTemporary(t)) {
                     if (LOGGER.isInfoEnabled()) {
                         LOGGER.info(
                             "[{}] Temporary error during tracking of upstream shard changes for subscription '{}'. Retrying: {}:{}",
@@ -263,6 +266,9 @@ public class ShardReplicationChangesTracker implements Closeable {
                                     long toSeqNoRequested,
                                     long toSeqNoReceived,
                                     long seqNoAtLeader) {
+        if (closed) {
+            return;
+        }
         if (success) {
             // we shouldn't ever be getting more operations than requested.
             assert toSeqNoRequested >= toSeqNoReceived :
@@ -318,8 +324,8 @@ public class ShardReplicationChangesTracker implements Closeable {
         // Renew retention lease with global checkpoint so that any shard that picks up shard replication task
         // has data until then.
         // Method is called inside a transport thread (response listener), so dispatch away
-        threadPool.executor(ThreadPool.Names.LOGICAL_REPLICATION).execute(() -> {
-            shardReplicationService.getRemoteClusterClient(shardId.getIndex()).thenAccept(remoteClient ->
+        threadPool.executor(ThreadPool.Names.LOGICAL_REPLICATION).execute(() ->
+            shardReplicationService.getRemoteClusterClient(shardId.getIndex(), subscriptionName).thenAccept(remoteClient ->
                 RetentionLeaseHelper.renewRetentionLease(
                     shardId,
                     toSeqNoReceived,
@@ -356,8 +362,7 @@ public class ShardReplicationChangesTracker implements Closeable {
                         }
                     )
                 )
-            );
-        });
+            ));
     }
 
     @Override
@@ -368,7 +373,7 @@ public class ShardReplicationChangesTracker implements Closeable {
             currentCancellable.cancel();
             cancellable = null;
         }
-        shardReplicationService.getRemoteClusterClient(shardId.getIndex())
+        shardReplicationService.getRemoteClusterClient(shardId.getIndex(), subscriptionName)
             .thenAccept(client -> RetentionLeaseHelper.attemptRetentionLeaseRemoval(
                 shardId,
                 clusterName,

--- a/server/src/main/java/io/crate/replication/logical/ShardReplicationService.java
+++ b/server/src/main/java/io/crate/replication/logical/ShardReplicationService.java
@@ -21,7 +21,16 @@
 
 package io.crate.replication.logical;
 
-import io.crate.replication.logical.exceptions.NoSubscriptionForIndexException;
+import static io.crate.replication.logical.LogicalReplicationSettings.REPLICATION_SUBSCRIPTION_NAME;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.annotation.Nullable;
+
 import org.apache.logging.log4j.Logger;
 import org.elasticsearch.client.Client;
 import org.elasticsearch.cluster.ClusterName;
@@ -29,25 +38,14 @@ import org.elasticsearch.cluster.service.ClusterService;
 import org.elasticsearch.common.inject.Inject;
 import org.elasticsearch.common.logging.Loggers;
 import org.elasticsearch.common.settings.Settings;
-import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
 import org.elasticsearch.index.Index;
-import org.elasticsearch.index.IndexService;
-import org.elasticsearch.index.IndexSettings;
 import org.elasticsearch.index.shard.IndexEventListener;
 import org.elasticsearch.index.shard.IndexShard;
 import org.elasticsearch.index.shard.ShardId;
-import org.elasticsearch.indices.cluster.IndicesClusterStateService.AllocatedIndices.IndexRemovalReason;
 import org.elasticsearch.threadpool.ThreadPool;
 import org.elasticsearch.transport.RemoteClusters;
 
-import javax.annotation.Nullable;
-import java.io.Closeable;
-import java.io.IOException;
-import java.util.Map;
-import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.ConcurrentHashMap;
-
-import static io.crate.replication.logical.LogicalReplicationSettings.REPLICATION_SUBSCRIPTION_NAME;
+import io.crate.replication.logical.exceptions.NoSubscriptionForIndexException;
 
 public class ShardReplicationService implements Closeable, IndexEventListener {
 
@@ -59,7 +57,6 @@ public class ShardReplicationService implements Closeable, IndexEventListener {
     private final ThreadPool threadPool;
     private final ClusterName clusterName;
     private final Map<ShardId, ShardReplicationChangesTracker> shards = new ConcurrentHashMap<>();
-    private final Map<String, String> subscribedIndices = ConcurrentCollections.newConcurrentMap();
     private final RemoteClusters remoteClusters;
 
     @Inject
@@ -84,11 +81,7 @@ public class ShardReplicationService implements Closeable, IndexEventListener {
         }
     }
 
-    CompletableFuture<Client> getRemoteClusterClient(Index index) {
-        String subscriptionName = subscribedIndices.get(index.getUUID());
-        if (subscriptionName == null) {
-            return CompletableFuture.failedFuture(new NoSubscriptionForIndexException(index));
-        }
+    CompletableFuture<Client> getRemoteClusterClient(Index index, String subscriptionName) {
         var subscription = logicalReplicationService.subscriptions().get(subscriptionName);
         if (subscription == null) {
             return CompletableFuture.failedFuture(new NoSubscriptionForIndexException(index));
@@ -97,27 +90,9 @@ public class ShardReplicationService implements Closeable, IndexEventListener {
     }
 
     @Override
-    public void afterIndexCreated(IndexService indexService) {
-        var settings = indexService.getIndexSettings().getSettings();
-        var subscriptionName = REPLICATION_SUBSCRIPTION_NAME.get(settings);
-        if (subscriptionName.isEmpty() == false) {
-            subscribedIndices.put(indexService.indexUUID(), subscriptionName);
-        }
-    }
-
-    @Override
-    public void afterIndexRemoved(Index index, IndexSettings indexSettings, IndexRemovalReason reason) {
-        var settings = indexSettings.getSettings();
-        var subscriptionName = REPLICATION_SUBSCRIPTION_NAME.get(settings);
-        if (subscriptionName.isEmpty() == false) {
-            subscribedIndices.remove(index.getUUID());
-        }
-    }
-
-    @Override
     public void afterIndexShardStarted(IndexShard indexShard) {
         if (indexShard.routingEntry().primary()) {
-            var subscriptionName = subscribedIndices.get(indexShard.shardId().getIndex().getUUID());
+            var subscriptionName = REPLICATION_SUBSCRIPTION_NAME.get(indexShard.indexSettings().getSettings());
             if (subscriptionName == null) {
                 return;
             }
@@ -156,7 +131,6 @@ public class ShardReplicationService implements Closeable, IndexEventListener {
             && REPLICATION_SUBSCRIPTION_NAME.get(newSettings).isEmpty() == true) {
             var shardId = indexShard.shardId();
             closeTracker(shardId);
-            subscribedIndices.remove(indexShard.shardId().getIndex().getUUID());
         }
     }
 

--- a/server/src/main/java/io/crate/replication/logical/action/ReplayChangesAction.java
+++ b/server/src/main/java/io/crate/replication/logical/action/ReplayChangesAction.java
@@ -44,6 +44,7 @@ import org.elasticsearch.common.inject.Singleton;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.common.lucene.uid.Versions;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.index.engine.Engine;
 import org.elasticsearch.index.mapper.StrictDynamicMappingException;
@@ -55,7 +56,10 @@ import org.elasticsearch.threadpool.ThreadPool;
 import org.elasticsearch.transport.TransportActionProxy;
 import org.elasticsearch.transport.TransportService;
 
+import io.crate.common.annotations.VisibleForTesting;
 import io.crate.common.unit.TimeValue;
+import io.crate.replication.logical.engine.SubscriberEngine;
+import io.crate.replication.logical.exceptions.InvalidShardEngineException;
 
 public class ReplayChangesAction extends ActionType<ReplicationResponse> {
 
@@ -68,8 +72,7 @@ public class ReplayChangesAction extends ActionType<ReplicationResponse> {
     }
 
     @Singleton
-    public static class TransportAction
-        extends TransportWriteAction<Request, Request, ReplicationResponse> {
+    public static class TransportAction extends TransportWriteAction<Request, Request, ReplicationResponse> {
 
         private static final Logger LOGGER = Loggers.getLogger(ReplayChangesAction.class);
 
@@ -101,18 +104,32 @@ public class ReplayChangesAction extends ActionType<ReplicationResponse> {
         }
 
         @Override
-        protected void shardOperationOnPrimary(Request request, IndexShard primary, ActionListener<PrimaryResult<Request, ReplicationResponse>> listener) {
-            performOnPrimary(primary, request, new ArrayList<>(), 0, results -> {
+        protected void shardOperationOnPrimary(Request request,
+                                               IndexShard primary,
+                                               ActionListener<PrimaryResult<Request, ReplicationResponse>> listener) {
+            performOnPrimary(primary, request, new ArrayList<>(),0, results -> {
+                ArrayList<Translog.Operation> replicaOps = new ArrayList<>();
                 Translog.Location location = null;
                 try {
-                    for (Engine.Result engineResult : results) {
+                    for (int i = 0; i < results.size(); i++) {
+                        var engineResult = results.get(i);
+                        var failure = engineResult.getFailure();
+                        if (failure instanceof InvalidShardEngineException) {
+                            if (location == null) {
+                                // Bail out if the failure happened on the first item -> nothing to replicate
+                                listener.onFailure(failure);
+                            }
+                            break;
+                        }
+                        // Add successfully on primary written operation to the to-replicate operations
+                        replicaOps.add(request.changes().get(i));
                         location = syncOperationResultOrThrow(engineResult, location);
                     }
                 } catch (Exception e) {
                     listener.onFailure(e);
                 }
                 listener.onResponse(new WritePrimaryResult<>(
-                    request,
+                    new Request(primary.shardId(), replicaOps, request.maxSeqNoOfUpdatesOrDeletes()),
                     new ReplicationResponse(),
                     location,
                     null,
@@ -121,13 +138,27 @@ public class ReplayChangesAction extends ActionType<ReplicationResponse> {
             }, listener::onFailure);
         }
 
-        private void performOnPrimary(IndexShard primary, Request request, List<Engine.Result> accumulator, int offset, Consumer<List<Engine.Result>> result, Consumer<Exception> failure) {
+        @VisibleForTesting
+        void performOnPrimary(IndexShard primary,
+                              Request request,
+                              List<Engine.Result> accumulator,
+                              int offset,
+                              Consumer<List<Engine.Result>> result,
+                              Consumer<Exception> failure) {
             for (int i = offset; i < request.changes().size(); i++) {
                 final var op = request.changes().get(i);
                 final var opWithPrimary = withPrimaryTerm(op, primary.getOperationPrimaryTerm());
                 if (primary.getMaxSeqNoOfUpdatesOrDeletes() < request.maxSeqNoOfUpdatesOrDeletes()) {
                     primary.advanceMaxSeqNoOfUpdatesOrDeletes(request.maxSeqNoOfUpdatesOrDeletes());
                 }
+
+                if ((primary.getEngineOrNull() instanceof SubscriberEngine) == false) {
+                    accumulator.add(
+                        new Engine.IndexResult(new InvalidShardEngineException(primary.shardId()), Versions.MATCH_ANY)
+                    );
+                    break;
+                }
+
                 Engine.Result engineResult = null;
                 try {
                     engineResult = primary.applyTranslogOperation(opWithPrimary, Engine.Operation.Origin.PRIMARY);
@@ -148,7 +179,12 @@ public class ReplayChangesAction extends ActionType<ReplicationResponse> {
             result.accept(accumulator);
         }
 
-        private void retryWhenMappingsAreUpdated(IndexShard primary, Request request, List<Engine.Result> accumulator, int offset, Consumer<List<Engine.Result>> result, Consumer<Exception> failure) {
+        private void retryWhenMappingsAreUpdated(IndexShard primary,
+                                                 Request request,
+                                                 List<Engine.Result> accumulator,
+                                                 int offset,
+                                                 Consumer<List<Engine.Result>> result,
+                                                 Consumer<Exception> failure) {
             var indexName = request.shardId().getIndexName();
             var currentMappingVersion = clusterService.state().metadata().index(indexName).getMappingVersion();
             var clusterStateObserver = new ClusterStateObserver(clusterService, LOGGER);

--- a/server/src/main/java/io/crate/replication/logical/exceptions/InvalidShardEngineException.java
+++ b/server/src/main/java/io/crate/replication/logical/exceptions/InvalidShardEngineException.java
@@ -1,0 +1,41 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.replication.logical.exceptions;
+
+import java.util.Locale;
+
+import org.elasticsearch.index.shard.ShardId;
+
+/**
+ * Marker exception raised if the engine of a target shard is not a
+ * {@link io.crate.replication.logical.engine.SubscriberEngine}.
+ * This is expected to happen if a subscription is dropped and an index is turned into a normal read-write
+ * engine while {@link io.crate.replication.logical.action.ReplayChangesAction} responses are still arriving.
+ */
+public class InvalidShardEngineException extends RuntimeException {
+
+    private static final String MESSAGE = "Shard '%s' is not subscribed anymore (wrong engine)";
+
+    public InvalidShardEngineException(ShardId shardId) {
+        super(String.format(Locale.ENGLISH, MESSAGE, shardId));
+    }
+}

--- a/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java
+++ b/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java
@@ -19,12 +19,41 @@
 
 package org.elasticsearch.index.shard;
 
-import com.carrotsearch.hppc.ObjectLongMap;
-import io.crate.common.Booleans;
-import io.crate.common.collections.Tuple;
-import io.crate.common.io.IOUtils;
-import io.crate.common.unit.TimeValue;
-import io.crate.exceptions.Exceptions;
+import static org.elasticsearch.index.seqno.SequenceNumbers.UNASSIGNED_SEQ_NO;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.io.UncheckedIOException;
+import java.nio.channels.ClosedByInterruptException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.LongSupplier;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+import javax.annotation.Nullable;
+
 import org.apache.logging.log4j.Logger;
 import org.apache.logging.log4j.message.ParameterizedMessage;
 import org.apache.lucene.index.CheckIndex;
@@ -126,39 +155,13 @@ import org.elasticsearch.repositories.Repository;
 import org.elasticsearch.rest.RestStatus;
 import org.elasticsearch.threadpool.ThreadPool;
 
-import javax.annotation.Nullable;
-import java.io.Closeable;
-import java.io.IOException;
-import java.io.PrintStream;
-import java.io.UncheckedIOException;
-import java.nio.channels.ClosedByInterruptException;
-import java.nio.charset.StandardCharsets;
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.EnumSet;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Locale;
-import java.util.Map;
-import java.util.Objects;
-import java.util.Optional;
-import java.util.Set;
-import java.util.concurrent.CopyOnWriteArrayList;
-import java.util.concurrent.CountDownLatch;
-import java.util.concurrent.TimeUnit;
-import java.util.concurrent.TimeoutException;
-import java.util.concurrent.atomic.AtomicBoolean;
-import java.util.concurrent.atomic.AtomicLong;
-import java.util.concurrent.atomic.AtomicReference;
-import java.util.function.BiConsumer;
-import java.util.function.Consumer;
-import java.util.function.Function;
-import java.util.function.LongSupplier;
-import java.util.stream.Collectors;
-import java.util.stream.StreamSupport;
+import com.carrotsearch.hppc.ObjectLongMap;
 
-import static org.elasticsearch.index.seqno.SequenceNumbers.UNASSIGNED_SEQ_NO;
+import io.crate.common.Booleans;
+import io.crate.common.collections.Tuple;
+import io.crate.common.io.IOUtils;
+import io.crate.common.unit.TimeValue;
+import io.crate.exceptions.Exceptions;
 
 public class IndexShard extends AbstractIndexShardComponent implements IndicesClusterStateService.Shard {
 
@@ -2454,7 +2457,7 @@ public class IndexShard extends AbstractIndexShardComponent implements IndicesCl
      * NOTE: returns null if engine is not yet started (e.g. recovery phase 1, copying over index files, is still running), or if engine is
      * closed.
      */
-    protected Engine getEngineOrNull() {
+    public Engine getEngineOrNull() {
         return this.currentEngineReference.get();
     }
 

--- a/server/src/test/java/io/crate/integrationtests/LogicalReplicationITest.java
+++ b/server/src/test/java/io/crate/integrationtests/LogicalReplicationITest.java
@@ -30,6 +30,8 @@ import java.util.ArrayList;
 import java.util.List;
 import java.util.Objects;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
 
 import org.elasticsearch.cluster.metadata.Metadata;
 import org.elasticsearch.cluster.service.ClusterService;
@@ -556,5 +558,64 @@ public class LogicalReplicationITest extends LogicalReplicationITestCase {
             OperationOnInaccessibleRelationException.class,
             "The relation \"doc.t1\" doesn't allow DROP operations, because it is included in a logical replication subscription."
         );
+    }
+
+    /**
+     * Test a regression which resulting in a failed initial restore of the re-created subscription caused
+     * not correctly removed retention leases while subscription is dropped.
+     */
+    @Test
+    public void test_drop_and_recreate_subscription_while_published_table_is_written_to() throws Exception {
+        executeOnPublisher("CREATE TABLE t1 (id INT) WITH(" + defaultTableSettings() +")");
+        executeOnPublisher("INSERT INTO t1 (id) VALUES (1), (2), (3), (4)");
+        createPublication("pub1", false, List.of("t1"));
+
+        var cnt = new AtomicLong(4L);
+        var stop = new AtomicBoolean(false);
+        try {
+            var writeThread = new Thread(() -> {
+                while (stop.get() == false) {
+                    publisherSqlExecutor.exec("INSERT INTO t1 (id) VALUES (?)", new Object[randomIntBetween(1, 9999)]);
+                    cnt.incrementAndGet();
+                    try {
+                        Thread.sleep(100);
+                    } catch (InterruptedException e) {
+                        stop.set(true);
+                    }
+                }
+            });
+            writeThread.start();
+
+            createSubscription("sub1", "pub1");
+
+            assertBusy(() -> {
+                var res = executeOnSubscriber(
+                    "SELECT" +
+                        " s.subname, r.relname, sr.srsubstate, sr.srsubstate_reason" +
+                        " FROM pg_subscription s" +
+                        " JOIN pg_subscription_rel sr ON s.oid = sr.srsubid" +
+                        " JOIN pg_class r ON sr.srrelid = r.oid");
+                assertThat(printedTable(res.rows()), Matchers.is("sub1| t1| r| NULL\n"));
+            });
+
+            executeOnSubscriber("DROP SUBSCRIPTION sub1");
+            executeOnSubscriber("DROP TABLE t1");
+
+            executeOnSubscriber("CREATE SUBSCRIPTION sub1" +
+                " CONNECTION '" + publisherConnectionUrl() + "' publication pub1");
+
+        } finally {
+            stop.set(true);
+        }
+
+        assertBusy(() -> {
+            var res = executeOnSubscriber(
+                "SELECT" +
+                    " s.subname, r.relname, sr.srsubstate, sr.srsubstate_reason" +
+                    " FROM pg_subscription s" +
+                    " JOIN pg_subscription_rel sr ON s.oid = sr.srsubid" +
+                    " JOIN pg_class r ON sr.srrelid = r.oid");
+            assertThat(printedTable(res.rows()), Matchers.is("sub1| t1| r| NULL\n"));
+        });
     }
 }

--- a/server/src/test/java/io/crate/replication/logical/action/ReplayChangesActionTest.java
+++ b/server/src/test/java/io/crate/replication/logical/action/ReplayChangesActionTest.java
@@ -1,0 +1,190 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.replication.logical.action;
+
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.nullValue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.replication.TransportWriteAction;
+import org.elasticsearch.cluster.action.shard.ShardStateAction;
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.IndexService;
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.translog.Translog;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.test.transport.MockTransportService;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+
+import io.crate.netty.NettyBootstrap;
+import io.crate.replication.logical.engine.SubscriberEngine;
+import io.crate.replication.logical.exceptions.InvalidShardEngineException;
+import io.crate.test.integration.CrateDummyClusterServiceUnitTest;
+
+public class ReplayChangesActionTest extends CrateDummyClusterServiceUnitTest {
+
+    private final static Index INDEX = new Index("index", UUIDs.randomBase64UUID());
+    protected final ShardId shardId = new ShardId(INDEX, 0);
+
+    private ReplayChangesAction.TransportAction transportAction;
+    private IndexShard indexShard;
+    private NettyBootstrap nettyBootstrap;
+
+
+    @Before
+    public void prepare() throws Exception {
+        nettyBootstrap = new NettyBootstrap(Settings.EMPTY);
+        nettyBootstrap.start();
+
+        IndicesService indicesService = mock(IndicesService.class);
+        IndexService indexService = mock(IndexService.class);
+
+        when(indicesService.indexServiceSafe(INDEX)).thenReturn(indexService);
+        indexShard = mock(IndexShard.class);
+        when(indexService.getShard(0)).thenReturn(indexShard);
+        when(indexShard.shardId()).thenReturn(shardId);
+
+        transportAction = new ReplayChangesAction.TransportAction(
+            Settings.EMPTY,
+            MockTransportService.createNewService(Settings.EMPTY, Version.CURRENT, THREAD_POOL, nettyBootstrap, clusterService.getClusterSettings()),
+            clusterService,
+            indicesService,
+            THREAD_POOL,
+            mock(ShardStateAction.class)
+        );
+    }
+
+    @After
+    public void tearDownNetty() throws Exception {
+        nettyBootstrap.close();
+    }
+
+    @Test
+    public void test_shardOperationOnPrimary_fails_if_engine_is_not_a_subscriber_engine_on_first_item() {
+        var req = new ReplayChangesAction.Request(
+            shardId,
+            List.of(
+                new Translog.NoOp(0, 0, "replicated"),
+                new Translog.NoOp(1, 0, "replicated")
+            ),
+            0
+        );
+        // Immediately fail because of wrong engine
+        when(indexShard.getEngineOrNull()).thenReturn(mock(Engine.class));
+
+        // Exception is already raised on first item, as no item is written yet, the whole request is expected to fail.
+
+        var exception = new AtomicReference<Throwable>();
+        transportAction.shardOperationOnPrimary(req, indexShard, ActionListener.wrap(r -> {}, exception::set));
+
+        assertThat(exception.get(), instanceOf(InvalidShardEngineException.class));
+    }
+
+    @Test
+    public void test_shardOperationOnPrimary_exclude_item_for_replication_if_failed_by_wrong_engine() throws Exception {
+        var req = new ReplayChangesAction.Request(
+            shardId,
+            List.of(
+                new Translog.NoOp(0, 0, "replicated"),
+                new Translog.NoOp(1, 0, "replicated")
+            ),
+            0
+        );
+
+        var engineResult = new Engine.IndexResult(0, 0, 0, true) {
+            @Override
+            public Translog.Location getTranslogLocation() {
+                return new Translog.Location(0, 0, 1);
+            }
+        };
+        when(indexShard.applyTranslogOperation(any(), any())).thenReturn(engineResult);
+        when(indexShard.getEngineOrNull()).thenAnswer(
+            new Answer<>() {
+                private int count = 0;
+                public Object answer(InvocationOnMock invocation) {
+                    if (count++ == 1) {
+                        return mock(Engine.class);
+                    }
+                    return mock(SubscriberEngine.class);
+                }
+            });
+
+        // Exception raised on second item, listener->failure must not be called as first item was written
+        // and must be replicated.
+        var result = new AtomicReference<TransportWriteAction.PrimaryResult>();
+        var exception = new AtomicReference<Throwable>();
+        transportAction.shardOperationOnPrimary(req, indexShard, ActionListener.wrap(result::set, exception::set));
+
+        assertThat(exception.get(), nullValue());
+        var replicaReq = (ReplayChangesAction.Request) result.get().replicaRequest();
+
+        // Failure happened on 2nd item, so only 1 item must be replicated
+        assertThat(replicaReq.changes().size(), is(1));
+    }
+
+    @Test
+    public void test_performOnPrimary_mark_result_as_failed_if_engine_is_not_a_subscriber_engine() throws Exception {
+        var req = new ReplayChangesAction.Request(
+            shardId,
+            List.of(
+                new Translog.NoOp(0, 0, "replicated"),
+                new Translog.NoOp(1, 0, "replicated")
+            ),
+            1
+        );
+        when(indexShard.applyTranslogOperation(any(), any())).thenReturn(new Engine.IndexResult(0, 0, 0, true));
+        when(indexShard.getEngineOrNull()).thenAnswer(
+            new Answer<>() {
+                private int count = 0;
+                public Object answer(InvocationOnMock invocation) {
+                    if (count++ == 1) {
+                        return mock(Engine.class);
+                    }
+                    return mock(SubscriberEngine.class);
+                }
+            });
+
+
+        ArrayList<Engine.Result> engineResults = new ArrayList<>();
+        transportAction.performOnPrimary(indexShard, req, new ArrayList<>(),0, engineResults::addAll, e -> {});
+
+        assertThat(engineResults.get(0).getFailure(), nullValue());
+        assertThat(engineResults.get(1).getFailure(), instanceOf(InvalidShardEngineException.class));
+    }
+}

